### PR TITLE
Dynamic volume gate

### PR DESCRIPTION
**Pull request by j3ffml**
Rebased on #31140, only review last commit.  Adds a feature-gate flag for dynamic volume provisioning alpha, defaulting to enabled to avoid breaking people. Key should be removed when support for the alpha version of this is removed.

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/31171)

### CODE REVIEW

File: pkg/util/config/feature_gate.go [hunk 1]
```diff
@@ -31,63 +31,104 @@ const (
...
+
+	// Every feature gate should add method here following this template:
+	//
+	// // owner: @username
+	// // alpha: v1.4
+	// MyFeature() bool
+
+	// DynamicVolumeProvisioningGate is a toggle for alpha dynamic volume
+	// provisioning support, which is controlled by annotations.
+	// owner: @saadali
```
Comment (saad-ali) [inline pkg/util/config/feature_gate.go]: `@saad-ali` (GitHub is different than internal)

**Comment by saad-ali** [comment]
LGTM but I'll do one more pass once #31140 merges, so it's easier to see the delta.

**Comment by jlowdermilk** [comment]
Rebased, ptal

**Comment by jlowdermilk** [comment]
Rebased on #30090. Fixed comments a bit.

File: pkg/util/config/feature_gate.go [hunk 2]
```diff
@@ -34,20 +34,29 @@ const (
...
 	//   AllAlpha=false,NewFeature=true  will result in newFeature=true
-	//   AllAlpha=true,NewFeature=false  will result in newFeature=false
-	allAlphaGate         = "AllAlpha"
-	dynamicKubeletConfig = "DynamicKubeletConfig"
+	//   NewFeature=false,AllAlpha=true will result in newFeature=false
+	allAlphaGate = "AllAlpha"
+
+	// DynamicKubeletConfig is a toggle for kubelet loading its
+	// config via a config map from the apiserver.
+	dynamicKubeletConfigGate = "DynamicKubeletConfig"
```
Comment (mtaufen) [inline pkg/util/config/feature_gate.go]: Is there a reason for naming the constant identifiers with the suffix `Gate`? I'm not sure I like the inconsistency of going from no-suffix string to suffixed identifier to no-suffix accessor.

**Comment by mtaufen** [comment]
+1 comment fixes

File: pkg/util/config/feature_gate.go (hunk 2 above)
Comment (jlowdermilk) [inline pkg/util/config/feature_gate.go]: A const with suffix `Gate` denotes a key in the `--feature-gates` map. I actually like this pattern, so keeping it unless more people complain.

File: pkg/util/config/feature_gate.go (hunk 1 above)
Comment (jlowdermilk) [inline pkg/util/config/feature_gate.go]: Fixed.

**Comment by jlowdermilk** [comment]
4th rebase, bumping priority.

**Pull request merged**